### ISSUE TITLE
Enum type not assignable to complete overloaded literal subsets

### ISSUE THREAD

**Issue by skeggse** [microsoft/pyright#5600]
**Describe the bug**

If you have a function with two different return types depending on which subset of the Enum it receives, you might model that behavior with `@overload`. The resulting function will not accept the full Enum value under pyright.

**Code or Screenshots**

```python
from enum import Enum
from typing import Literal, overload

class MyEnum(Enum):
    A = 1
    B = 2
    C = 3

@overload
def bip(value: Literal[MyEnum.A]) -> int:
    ...

@overload
def bip(value: Literal[MyEnum.B, MyEnum.C]) -> int:
    ...

def bip(value: MyEnum) -> int:
    return value.value

def outer(value: MyEnum):
    # This should work, but it does not.
    return bip(value)

def outer_alt1(value: Literal[MyEnum.A, MyEnum.B, MyEnum.C]):
    return bip(value)

def outer_alt2(value: MyEnum):
    if value == MyEnum.A:
        return bip(value)
    return bip(value)
```

Results in an error for one of the "outer" functions.

```
repro.py
repro.py:27:16 - error: Argument of type "MyEnum" cannot be assigned to parameter "value" of type "Literal[MyEnum.B, MyEnum.C]" in function "bip"
    Type "MyEnum" cannot be assigned to type "Literal[MyEnum.B, MyEnum.C]"
      "MyEnum" cannot be assigned to type "Literal[MyEnum.B]"
      "MyEnum" cannot be assigned to type "Literal[MyEnum.C]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Command-line tool (`v1.1.318`), but also VS Code extension.

**Comment by erictraut** [comment]
This is by design, not a bug. When pyright is matching overloads, it will perform "union expansion", but it doesn't perform expansion of enum types. Neither do other Python type checkers like mypy. Doing so would be costly from a performance standpoint. 

If you want this to work in your code, you will need to manually expand the enum into it constituent literal types.

```python
MyEnumValues = Literal[MyEnum.A] | Literal[MyEnum.B] | Literal[MyEnum.C]

def outer(value: MyEnumValues):
    return bip(value)
```
